Add `evalSystem` and `evalPackages` project args

hamishmack opened the pull request:
This adds a way to specify the `evalSystem` or `evalPackages` explicitly when calling the `project` functions.

Currently if we want to make a `flake` that supports multiple systems we have few options:

* Require builders for all the supported systems (even just for `nix flake show`).

* Pass `--impure` so that haskell.nix can see `builtins.currentSystem` to set up `pkgs.evalPackages` to use that.  Unfortunately this prevents nix from caching some of the work it does and often results in it recalculating for each supported system when it would otherwise be cached and take no time at all.

* Add an overlay to replace `evalPackages`.  This works, but it is not straight forward.

* Materialize the nix files for the project.

This change allows `evalSystem = "x86_64-linux";` to be passed telling `haskell.nix` to run `cabal` and `nix-tools` on that system.  The user will have to have a builder for that system, but does not need to have builders for the others (unless building outputs for them).

michaelpj commented:
I wonder if it would make sense to ask the user to provide `evalPackages` instead? The only way we use `evalSystem` is to get `evalPackages`, and it means we have to do the awful dance with re-importing nixpkgs by getting the original path and everything, whereas the user should just be able to re-import their nixpkgs straightforwardly, so it makes things a little less evil. WDYT?

michaelpj commented on lib/call-stack-to-nix.nix:
```diff
@@ -15,9 +15,16 @@
+, ... }@args:
 let
+  pkgs = defaults.pkgs.haskell-nix.evalPackagesFor evalSystem;
+  nix-tools = if args.nix-tools != null
```
This seems worse to me...

michaelpj commented on lib/call-stack-to-nix.nix:
```diff
@@ -15,9 +15,16 @@
+, evalSystem ? null
+, ... }@args:
 let
+  pkgs = defaults.pkgs.haskell-nix.evalPackagesFor evalSystem;
```
can we call this `evalPkgs`? I think it's really helpful to keep the distinction obvious, in fact currently this is shadowing the `pkgs` from line 8, which seems suspicious?

michaelpj commented on lib/stack-cache-generator.nix:
```diff
@@ -17,10 +17,18 @@
-}:
+}@args:
 let
+    pkgs = defaults.pkgs.haskell-nix.evalPackagesFor evalSystem;
```
ditto

michaelpj commented on modules/project-common.nix:
```diff
@@ -24,6 +24,18 @@ with lib.types;
+      default = null;
+      description = ''
+        Specifies the system on which `cabal configure` and `plan-to-nix` should run.
+        If not specified the `evalPackages` default will be used (builtins.currentSystem or
```
`evalPackages` is our internal jargon, can we say something more comprehensible to a user?

hamishmack replied:
I'm not too keen on it either, but I can't think of a better name (only other name I could think of was `bootPackages`).

michaelpj commented on modules/stack-project.nix:
```diff
@@ -42,7 +42,7 @@ with types;
     nix-tools = mkOption {
       type = nullOr package;
-      default = pkgs.evalPackages.haskell-nix.internal-nix-tools;
+      default = null;
```
ah, I see why the change

michaelpj commented on overlays/eval-packages.nix:
```diff
@@ -1,14 +1,23 @@
   # derivations that use `evalPackages` (these should be eval time only derviations
   # used to build nix inputs for IFD, the generated nix should match and so derivations
   # that depend on the IFD should match).
   evalPackages = import final.path {
```
Can't we just get rid of this now we're passing it everywhere?

hamishmack replied:
It's gone

michaelpj commented on overlays/haskell.nix:
```diff
@@ -276,9 +276,7 @@ final: prev: {
-            pkgs = final.evalPackages.pkgs;
-            inherit (final.evalPackages.pkgs) runCommand;
-            inherit (final.evalPackages.haskell-nix) mkCacheFile materialize haskellLib;
+            pkgs = final;
```
suspicious?

michaelpj replied:
I guess this does all line up: you pass in `pkgs` which will be used only to get the `evalPkgs`.

hamishmack commented:
I have removed `evalPackages` from `pkgs` and replaced it with an `evalPackages` arg.  I had to refactor `hackage-project` (and the functions that use it) to be nix modules based (so that the `nix-tools` argument could be passed correctly to match `evalPackages`).  The `hackage-quirks` are now a list of project level nix modules used by `hackage-package`.  

I removed `caller` arg as it was not actually used and I think `name` is more useful to have in error messages anyway.

I left `evalSystem` in, but it is just used in the default for `evalPackages` now.

michaelpj commented on lib/default.nix:
```diff
@@ -202,7 +199,7 @@ in {
     inherit lib cleanSourceWith;
     git = gitMinimal;
-    inherit (pkgs.evalPackages.buildPackages) runCommand;
+    inherit (pkgs.buildPackages.buildPackages) runCommand;
```
Are these changes actually okay? Won't this again mean that you can't eval without a build machine of the right system, because it will try and clean the source on a build machine?

hamishmack replied:
I think there is a situation when not using flakes (so `cleanGit` would have work to do) and cross compiling where this would have been useful so I have included a note in the change log.

michaelpj commented on lib/default.nix:
```diff
@@ -260,17 +257,19 @@ in {
   # Find the resolver in the stack.yaml file and fetch it if a sha256 value is provided
   fetchResolver = import ./fetch-resolver.nix {
-    inherit (pkgs.evalPackages) pkgs;
+    inherit (pkgs.buildPackages) pkgs;
```
similarly

hamishmack replied:
`fetchResolver` is only called from `lib/stack-cache-generator.nix` and that only has access to `evalPackages` now (does not have `pkgs` at all).

michaelpj commented on modules/project-common.nix:
```diff
@@ -24,6 +24,31 @@ with lib.types;
+        else if pkgs.buildPackages.system == config.evalSystem
+          then pkgs.buildPackages
+        else
+          import pkgs.path {
```
could also throw in this case and require the user to specify?

hamishmack replied:
I've changed the default to use `buildPackages`.  I also removed the default to make sure nothing in `haskell.nix` itself (including the CI) depends on the default.  The `import` code here should now only get used if `evalSystem` was explicitly passed in (for instance if the user passed `evalSystem = builtins.currentSystem;`).

michaelpj reviewed (commented):
Maybe this was a bad idea, I don't know. It seems like we now don't have `evalPackages` in the places where we need them?

hamishmack commented on lib/system-pkgs.nix:
```diff
@@ -12,10 +12,10 @@ pkgs:
   # if it comes from the targetPackages we won't even
   # be able to execute it.
-  // { fetchgit = pkgs.evalPackages.fetchgit; }
+  // { fetchgit = pkgs.buildPackages.fetchgit; }
```
I checked and it does look like the problem with `gitMinimal` has been fixed in nixpkgs 22.05.  I think we should keep this fix for another year or so for people using older `nixpkgs`.  I've moved this to `overlays/nix-prefetch-git-minimal.nix`.  I also included a better comment and the `nix-diff` command I used for checking `nixpkgs`.

hamishmack replied:
This turned out to be a small [bug in the libredirect derivation in nixpkgs](https://github.com/NixOS/nixpkgs/pull/182143).  I've put an override of libredirect in the overlay instead of messing with `fetchgit`.

hamishmack commented on overlays/tools.nix:
```diff
@@ -57,50 +57,31 @@ in { haskell-nix = prev.haskell-nix // {
-          // { name = final.haskell-nix.toolPackageName.${name} or name; }))
-          .getComponent "exe:${final.haskell-nix.packageToolName.${name} or name}";
+      package = final.haskell-nix.hackage-package (
+        (if builtins.isList projectModule then projectModule else [projectModule])
```
I have changed `hackage-tool` so it only takes a single module.  If there is a way to compose nix modules we could change the others too (`hackage-package`, `hackage-project`, `cabalProject`, etc.).  Alternatively we could also split out variants that take a list and have wrappers for those that take a single module.  Let me know if you think that would be nicer.

michaelpj replied:
I think we suffer a lot from not having types, so I guess the heuristic I end up going for a lot is that an argument named `X` should always have the same type. So I would just use a list of modules everywhere just so it's super consistent, even if that's a bit more annoying in places where you probably only want one?

hamishmack replied:
I think it might make sense to move `cabalProject` and `stackProject` code from `overlays/haskell.nix` to `modules/cabal-project.nix` and `overlays/stack-project.nix`.  I think then we might be able to simplify the way we eval the project modules.   However I started trying that and it seems like a much bigger change.  So for now I have changed `hackage-tool` to accept only a list.

hamishmack commented:
bors try

hamishmack commented:
bors try

dhess commented:
I believe this commit broke the evaluation of our `haskell.nix` project with the following error:

```
error: value is the partially applied built-in function 'mapAttrs' while a set was expected
(use '--show-trace' to show detailed location information)
```

I've tracked it down to this commit by pinning our `haskell.nix` Flake input to 7f02c057e6820624d878968e1e7bd11c3e295cde; after that commit, this error occurs during evaluation.

hamishmack commented:
Should be fixed by #1570

dhess commented:
> Should be fixed by #1570

Thanks! I can confirm that it's fixed now.
